c++/error-handling.md

```diff
@@ -0,0 +1,981 @@
+{
+  return char(number - min_number + 'A');
+}
+```
+
+It so happens that we can see the definition of `number_to_letter`. Even if we couldn't,
+we might feel confident guessing roughly how it was implemented.
+And (assertions aside) we might feel confident to say that the function does not
+exhibit undefined behaviour, even when the precondition is violated.
+
+We would be committing a grave error.
+
+Firstly, the author has effectively limited the warranty of the contract.
+They are declining to make a guarantee about what the result of the call will be
+in the fact of contract violation. Put another way, they decline to define the behaviour.
+
+Secondly, the provider is under no obligation
+to have tested the behaviour outside of contract.
+So any and all assurances gathered through automated testing
+do not apply to out-of-contract use.
+And in the absence of testing, it's easy to miss a failure case, e.g.:
+
+```c++
+// signed integer overflow resulting in ISO C++ Standard UB
+number_to_letter(0x7fffffff);
+```
+
+Finally, the provider reserves the right to change the implementation as they choose.
+Assumptions drawn from the original definition are already false.
+But they become yet more dangerous if applied to a new definition, for example:
+
+```c++
+constexpr auto number_to_letter(int number)
+{
+  constexpr auto lookup_table = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+  return lookup_table[number - min_number];
+}
+```
+
+It is easy to become distracted by the wonders of modern compilers.
+They are able to generate highly optimised code, by assuming contract fulfilment.
+But at the end of the day, they are not magic and nor is undefined behaviour.
+UB is just a way for authors to limit contracts so that providers can deliver more.
+
+### Don't Optimize Until You Sanitize
+
+C++ toolchains increasingly optimise programs assuming that they are correct.
+Optimisations are not free. They demand strict adherence to contracts.
+Historically, UB has been difficult to detect.
+
+Fortunately, toolchains now also provide facilities for detecting some UB.
+These include dynamic analysis tools known as sanitizers
+which instrument code to test for bugs.
+The beauty of undefined behaviour is that it allows this -- and every other valid
+bug-hunting tool -- to be used in conforming code.
+
+Meanwhile, modern software development practices emphasise automated testing regimes.
+It is not uncomment for most of the APIs in a program
+to be tested for provider C++ API Contract violation frequently during development.
+Therefore, there is strong reason to use sanitizers as a matter of good practice.
+
+If you wish to enable compiler optimisations in the program you provide to users
+you are strongly advised to test your program code with sanitizers.
+Combined with the approaches detained in this document, this will lead to significantly
+less defective software without the necessity to compromise on safety or performance.
+
+### Don't Rely on Sanitizers Alone
```

Comment:
Follow up with a call to tool chain implementers to enable more warnings by default and enable sanitizers by default in -O0. Discuss legacy projects and the harm they do for new projects.